python/ray/rpdb.py

```diff
@@ -0,0 +1,181 @@
+# Some code in this file is from
+# https://github.com/ionelmc/python-remote-pdb/blob/07d563331c4ab9eb45731bb272b158816d98236e/src/remote_pdb.py
+# (BSD 2-Clause "Simplified" License)
+
+import errno
+import inspect
+import json
+import logging
+import os
+import re
+import socket
+import sys
+import uuid
+from pdb import Pdb
+import setproctitle
+import traceback
+
+from ray.experimental.internal_kv import _internal_kv_del, _internal_kv_put
+
+PY3 = sys.version_info[0] == 3
+log = logging.getLogger(__name__)
+
+
+def cry(message, stderr=sys.__stderr__):
+    log.critical(message)
+    print(message, file=stderr)
+    stderr.flush()
+
+
+class LF2CRLF_FileWrapper(object):
+    def __init__(self, connection):
+        self.connection = connection
+        self.stream = fh = connection.makefile("rw")
+        self.read = fh.read
+        self.readline = fh.readline
+        self.readlines = fh.readlines
+        self.close = fh.close
+        self.flush = fh.flush
+        self.fileno = fh.fileno
+        if hasattr(fh, "encoding"):
+            self._send = lambda data: connection.sendall(
+                data.encode(fh.encoding))
+        else:
+            self._send = connection.sendall
+
+    @property
+    def encoding(self):
+        return self.stream.encoding
+
+    def __iter__(self):
+        return self.stream.__iter__()
+
+    def write(self, data, nl_rex=re.compile("\r?\n")):
+        data = nl_rex.sub("\r\n", data)
+        self._send(data)
+
+    def writelines(self, lines, nl_rex=re.compile("\r?\n")):
+        for line in lines:
+            self.write(line, nl_rex)
+
+
+class RemotePdb(Pdb):
+    """
+    This will run pdb as a ephemeral telnet service. Once you connect no one
+    else can connect. On construction this object will block execution till a
+    client has connected.
+    Based on https://github.com/tamentis/rpdb I think ...
+    To use this::
+        RemotePdb(host="0.0.0.0", port=4444).set_trace()
+    Then run: telnet 127.0.0.1 4444
+    """
+    active_instance = None
+
+    def __init__(self, host, port, patch_stdstreams=False, quiet=False):
+        self._quiet = quiet
+        self._patch_stdstreams = patch_stdstreams
+        self._listen_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._listen_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR,
+                                       True)
+        self._listen_socket.bind((host, port))
+
+    def listen(self):
+        if not self._quiet:
+            cry("RemotePdb session open at %s:%s, waiting for connection ..." %
```

Comment:
Can you also include "use `ray debug`" to connect? It's not intuitive you have to do that.